src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,702 @@
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.MoveDeclarationNearReference;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Simplification;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.RemoveUnusedParametersAndValues
+{
+    /// <summary>
+    /// Code fixer for unused expression value diagnostics reported by <see cref="AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer"/>.
+    /// We provide following code fixes:
+    ///     1. If the unused value assigned to a local/parameter has no side-effects,
+    ///        we recommend removing the assignment. We consider an expression value to have no side effects
+    ///        if one of the following is true:
+    ///         1. Value is a compile time constant.
+    ///         2. Value is a local or parameter reference.
+    ///         3. Value is a field reference with no or implicit this instance.
+    ///     2. Otherwise, if user preference is set to DiscardVariable, and project's
+    ///        language version supports discard variable, we recommend assigning the value to discard.
+    ///     3. Otherwise, we recommend assigning the value to a new unused local variable which has no reads.
+    /// </summary>
+    internal abstract class AbstractRemoveUnusedParametersAndValuesCodeFixProvider<TExpressionSyntax, TStatementSyntax, TBlockSyntax,
+                                                                                   TExpressionStatementSyntax, TLocalDeclarationStatementSyntax,
+                                                                                   TVariableDeclaratorSyntax, TForEachStatementSyntax,
+                                                                                   TSwitchCaseBlockSyntax, TSwitchCaseLabelOrClauseSyntax,
+                                                                                   TCatchStatementSyntax, TCatchBlockSyntax>
+        : SyntaxEditorBasedCodeFixProvider
+        where TExpressionSyntax : SyntaxNode
+        where TStatementSyntax : SyntaxNode
+        where TBlockSyntax : TStatementSyntax
+        where TExpressionStatementSyntax : TStatementSyntax
+        where TLocalDeclarationStatementSyntax : TStatementSyntax
+        where TVariableDeclaratorSyntax : SyntaxNode
+        where TForEachStatementSyntax: TStatementSyntax
+        where TSwitchCaseBlockSyntax: SyntaxNode
+        where TSwitchCaseLabelOrClauseSyntax: SyntaxNode
```

Comment:
nit: when we've had these in the past, we've tried to group them by inheritance.  so, for example, all the things that are `: TStatementSyntax` would be grouped together.